File: 2022/07/p.js
#!/usr/bin/env node
const fs = require('fs');
const input = fs.readFileSync(0).toString();

class Node {
    constructor(parent = null) {
        this.parent = parent;
        this.dirs = {};
        this.files = {};
    }

    cd(name) {
        if (this.dirs[name] === undefined) {
            this.dirs[name] = new Node(this);
        }
        return this.dirs[name];
    }

    *visit() {
        this.total = Object.values(this.files).reduce((a,b) => a+b, 0);
        for (const dir of Object.values(this.dirs)) {
            yield* dir.visit();
            this.total += dir.total;
        }
        yield this;
    }
};

const root = new Node();
let cur = root;
for (let line of input.trim().split('\n')) {
    let tok = line.split(' ');
    if (tok[0] == '$') {
        if (tok[1] == 'cd') {
            if (tok[2] == '/') {
                cur = root;
            } else if (tok[2] == '..') {
                cur = cur.parent;
            } else {
                cur = cur.cd(tok[2]);
            }
        }
    } else if (tok[0] == 'dir') {
        cur.cd(tok[1]);
    } else {
        cur.files[tok[1]] = parseInt(tok[0]);
    }
}

const dirs = [...root.visit()];
let ans1 = 0, ans2 = 10**100;
for (const dir of dirs) {
    if (dir.total <= 100000) {
        ans1 += dir.total;
    }
    let f = 70000000 - root.total + dir.total;
    if (f >= 30000000) {
        ans2 = Math.min(ans2, dir.total);
    }
}
console.log(ans1, ans2);

File: 2022/07/p.py
#!/usr/bin/env python
import sys
from pathlib import Path

class Node:
    def __init__(self, parent, path):
        self.parent = parent
        self.path = path
        self.dirs = {}
        self.files = {}

    def subdir(self, name):
        if name not in self.dirs:
            self.dirs[name] = Node(self, self.path / name)
        return self.dirs[name]

    def rec(self):
        self.total = sum(self.files.values())
        for d in self.dirs.values():
            yield from d.rec()
            self.total += d.total
        yield self


root = cur = Node(None, Path('/'))
for line in sys.stdin:
    tok = line.strip().split()
    if tok[0] == '$':
        if tok[1] == 'cd':
            if tok[2] == '/':
                cur = root
            elif tok[2] == '..':
                cur = cur.parent
            else:
                cur = cur.subdir(tok[2])
    elif tok[0] == 'dir':
        cur.subdir(tok[1])
    else:
        cur.files[tok[1]] = int(tok[0])

all_dirs = list(root.rec())
ans1 = sum(d.total for d in all_dirs if d.total <= 100000)
ans2 = root.total + 1
for d in all_dirs:
    f = 70000000 - root.total + d.total
    if f >= 30000000:
        ans2 = min(ans2, d.total)
print(ans1, ans2)
